feat: integrate swap+send into existing tx flow

BZahory opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Swap+Send is currently decoupled from both the STX flow (we only use traditional transactions with swap+send) and the confirmations flow (we bypass this all together).

This PR reintegrates swap+send transactions with the existing STX and confirmation flows; it will also:
- Disable "edit" for swap+send transactions
- Remove the CTA button and tooltip previously used
- Update the copy on the approvals page
- Hide the failing simulation when the approval is still pending

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/25734?quickstart=1)

## **Related issues**

Resolves: [METABRIDGE-904](https://consensyssoftware.atlassian.net/browse/METABRIDGE-904)

## **Manual testing steps**

### General Flow

1. Enter the send flow on any supported swap+send network (e.g., BNB)
2. Fetch a swap+send quote from an ERC-20
3. Ensure that the "Continue" button does not submit a transaction on click or show a tooltip on hover
4. Click "Continue"
5. Ensure you can't edit the transaction
6. Ensure that the correct title copy is shown for that asset (e.g., "APPROVE USDC FOR SWAPS")
7. Submit the approval
8. Ensure that you progress to the trade transaction
9. Ensure that you can't edit the transaction
10. Ensure that you do not get a notification saying the transaction might fail
11. Ensure that the transaction can be submitted and funds are received by the recipient

### Smart Transaction Flow
1. Go to Ethereum Mainnet
2. Enter the send flow
3. Get a quote for ERC-20 -> any different asset
4. Click "Continue"
5. Confirm the approval transaction
6. Wait for the STX to complete; ensure you cannot submit the trade until it's done
7. Submit the trade and wait for the STX to complete
8. Ensure that the transaction is submitted and funds are received by the recipient

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

https://github.com/MetaMask/metamask-extension/assets/44588480/df60cb50-297d-4ad5-b60d-81a8ef2ff542


<!-- [screenshots/recordings] -->

### **After**

https://github.com/MetaMask/metamask-extension/assets/44588480/29a4ae65-4dae-4863-bf65-cffd18825f4f

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

Commit:
chore!: remove stx bypass

Commit:
feat: route tx through confirmations page

Commit:
fix: pass source token to copy

Commit:
chore: hide simulated changes when pending approval

Commit:
chore: remove CTA button logic

Commit:
chore: hide edit button for swap+send tx's

Commit:
fix: add prod endpoint back

Commit:
style: remove unused copy

Commit:
test: remove tooltip from snapshot

Commit:
test: update action

Commit:
test: update submission flow

Commit:
test: test for hidden simulation results

micaelae commented on ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -522,9 +523,14 @@ export default class ConfirmTransactionBase extends Component {
 
-    const simulationDetails = (
+    // if transaction reverts in simulation and there's an approval transaction, we don't want to show the simulation details
+    const isUnsimulable =
+      approvalTxId &&
+      simulationData?.error.code === SimulationErrorCode.Reverted;
```
Does the tx get re-simulated after the approval tx is confirmed?

BZahory replied:
Not that I'm aware of, but we set `isUnsimulable` to `true` if the simulation errors with a revert

Commit:
test: add tests for hidden edit button

remove unneeded literals

Commit:
test: add button click for approval

09306677806 commented:
0x88C478836e2f65b0322fdCB4e7f6A3e685c09e99

در تاریخ چهارشنبه ۱۰ ژوئیه ۲۰۲۴،‏ ۲۳:۴۷ sonarcloud[bot] <
***@***.***> نوشت:

> [image: Quality Gate Passed]
> <https://sonarcloud.io/dashboard?id=metamask-extension-private&pullRequest=25734> *Quality
> Gate passed*
>
> Issues
>
> <https://camo.githubusercontent.com/da77210611fad32e47ac6a36190d5e5320fe19b1c6006fae418d539fadaadc93/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67> 2
> New issues
> <https://sonarcloud.io/project/issues?id=metamask-extension-private&pullRequest=25734&resolved=false&sinceLeakPeriod=true>
>
> <https://camo.githubusercontent.com/6e7f22e3d7f9764d9cfb28e3b8ee5ebb5f711c5bb2dcec8a6db6b0d958d63134/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f61636365707465642d313670782e706e67> 0
> Accepted issues
> <https://sonarcloud.io/project/issues?id=metamask-extension-private&pullRequest=25734&resolutions=WONTFIX>
>
> Measures
>
> <https://camo.githubusercontent.com/da77210611fad32e47ac6a36190d5e5320fe19b1c6006fae418d539fadaadc93/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67> 0
> Security Hotspots
> <https://sonarcloud.io/project/security_hotspots?id=metamask-extension-private&pullRequest=25734&resolved=false&sinceLeakPeriod=true>
>
> <https://camo.githubusercontent.com/da77210611fad32e47ac6a36190d5e5320fe19b1c6006fae418d539fadaadc93/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67> 0.0%
> Coverage on New Code
> <https://sonarcloud.io/component_measures?id=metamask-extension-private&pullRequest=25734&metric=new_coverage&view=list>
>
> <https://camo.githubusercontent.com/da77210611fad32e47ac6a36190d5e5320fe19b1c6006fae418d539fadaadc93/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f7061737365642d313670782e706e67> 0.0%
> Duplication on New Code
> <https://sonarcloud.io/component_measures?id=metamask-extension-private&pullRequest=25734&metric=new_duplicated_lines_density&view=list>
>
> See analysis details on SonarCloud
> <https://sonarcloud.io/dashboard?id=metamask-extension-private&pullRequest=25734>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MetaMask/metamask-extension/pull/25734#issuecomment-2221360338>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANEZZRCYHB2N74GPNFMLVXDZLWJERAVCNFSM6AAAAABKUNUANOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEMRRGM3DAMZTHA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

Commit:
test: wait for approval to submit before clicking confirm

bschorchit commented:
For visibility, there's some [ongoing discussion](https://consensys.slack.com/archives/C068SFX90PN/p1720702160867999?thread_ts=1720616402.512989&cid=C068SFX90PN) on the reasoning for introducing the 2 confirmations instead of a single custom one.

matthewwalsh0 commented on ui/ducks/send/send.js:
```diff
@@ -3013,35 +3012,36 @@ export function signTransaction(history) {
-                  type: TransactionType.swapApproval,
-                  sourceTokenSymbol,
+          const { id } = await dispatch(
+            addTransactionAndRouteToConfirmationPage(
```
I want to stress I think this PR is a great improvement, since we're now centralising the transaction flows through a consistent code path, confirmation, and UX.

But as a point of interest and future discussion, the send flow architecture, including the associated duck, Redux state, and draft transaction, all introduce a lot of complexity on top of the transaction code paths that can feel more a side effect of legacy code as opposed to intentional design.

If we were to implement wallet transfers for the first time today, it feels it would be far simpler, modular, and maintainable, if we were to think of a transfer as fundamentally just a standard transaction confirmation but with some additional modular React components to specify tokens and networks and amounts, and update the underlying data.

As a result, we could replace lots of the large send logic like this function, with many modular, typed and functional React components and hooks, and use them to update a pending transaction directly in the `TransactionController` state. This would mean the existing Redux state could be replaced by the React components state and the `TransactionController` unapproved transaction.

This would also align with the ongoing confirmation re-design which is based around a single confirmation template with variations, custom inputs and rows dependent on the type.

matthewwalsh0 commented on ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -522,9 +523,14 @@ export default class ConfirmTransactionBase extends Component {
 
-    const simulationDetails = (
+    // if transaction reverts in simulation and there's an approval transaction, we don't want to show the simulation details
+    const isUnsimulable =
```
```suggestion
    const isUnsimulatable =
```

matthewwalsh0 commented on ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -522,9 +523,14 @@ export default class ConfirmTransactionBase extends Component {
+      approvalTxId &&
+      simulationData?.error.code === SimulationErrorCode.Reverted;
+
+    const simulationDetails = isUnsimulable ? undefined : (
```
Given the complexity and many overlapping scenarios, the `SimulationDetails` component generates its own error messages based on the result.

Why is this a special case in which the error is not shown? Is it desirable to introduce that inconsistency?

If we definitely want this, could we instead provide a new property such as `hideErrors` so we can keep the error message logic defined in one place within the component?

matthewwalsh0 commented on ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -1152,7 +1171,11 @@ export default class ConfirmTransactionBase extends Component {
           toEns={toEns}
           toNickname={toNickname}
-          showEdit={!isContractInteractionFromDapp && Boolean(onEdit)}
+          showEdit={
```
Minor, but as this grows, does it warrant a `showEdit` boolean?

matthewwalsh0 commented on ui/pages/confirmations/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -1129,14 +1135,27 @@ export default class ConfirmTransactionBase extends Component {
       (isTokenApproval || isContractInteraction) &&
       txData.origin !== 'metamask';
+
+    // TODO: swapApproval is exclusively used for swap+send currently,
```
To clarify, we use this in both `ui/ducks/send/send.js` and `ui/ducks/swaps/swaps.js`?

matthewwalsh0 commented on ui/store/actions.ts:
```diff
@@ -976,16 +976,20 @@ export function updateTransaction(
- * @param options
- * @param options.sendFlowHistory - The history of the send flow at time of creation.
- * @param options.type - The type of the transaction being added.
+ * @param txParams - the transaction parameters
```
```suggestion
 * @param txParams - The transaction parameters
```

matthewwalsh0 commented on ui/store/actions.ts:
```diff
@@ -976,16 +976,20 @@ export function updateTransaction(
+ * @param options.swaps - Options specific to swaps transactions.
+ * @param options.swaps.hasApproveTx - Whether the swap required an approval transaction.
+ * @param options.swaps.meta - Additional transaction metadata required by swaps.
+ * @param options.type
```
```suggestion
 * @param options.type - The type of the transaction being added.
```

matthewwalsh0 commented on ui/helpers/utils/transactions.util.js:
```diff
@@ -183,9 +183,15 @@ export function getStatusKey(transaction) {
+  t,
+  type,
+  nativeCurrency = 'ETH',
+  sourceToken,
```
Minor, should we default this to `undefined` to make it clear it's optional?

matthewwalsh0 commented on ui/ducks/send/send.js:
```diff
@@ -3013,35 +3012,36 @@ export function signTransaction(history) {
-            requireApproval: false,
+        const { id: swapAndSendTxId } = await dispatch(
+          addTransactionAndRouteToConfirmationPage(txParams, {
+            requireApproval: true,
```
Since we're now showing the confirmation, do we also want to run Blockaid validation as normal?

So should we update `SECURITY_PROVIDER_EXCLUDED_TRANSACTION_TYPES` in `shared/constants/security-provider.ts`?

Commit:
chore: use `swapAndSendApproval` type

Commit:
chore: accept approval params for stx

Commit:
chore: impl `submitSwapAndSendApprovalAndTrade`

Commit:
feat: process swap+send approvals+trades in one confirmation screen

Commit:
temp!: see body

You need to link the branch in https://github.com/MetaMask/core/pull/4634, so build and replace my name w your name, then once matt walsh is done yapping it up, get that merged and deployed, then finally replace all the linking with the latest npm version for the user-operations-controller and transaction-controller.
